soccer/gameplay/evaluation/double_touch.py

```diff
@@ -33,66 +36,105 @@ def __init__(self):
         self.add_transition(
             DoubleTouchTracker.State.start,
             DoubleTouchTracker.State.restart_play_began,
-            lambda: (main.root_play().play != None and main.root_play().play.__class__.is_restart() and main.game_state().is_our_restart()),
+            lambda: (main.root_play().play is not None and
+                     main.root_play().play.__class__.is_restart() and
+                     main.game_state().is_our_restart()),
             'we start running an offensive restart play')
 
         self.add_transition(
             DoubleTouchTracker.State.restart_play_began,
             DoubleTouchTracker.State.kicking,
-            lambda: (any(bot.has_ball() for bot in main.our_robots())) and not main.game_state().is_placement(),
-            'one of our bots has the ball')
+            lambda: ((any(self._has_ball(bot) for bot in main.our_robots())) or
+                     main.game_state().is_playing()) and not main.game_state().is_placement(),
+            'one of our bots has the ball or the ball was kicked')
 
         self.add_transition(DoubleTouchTracker.State.kicking,
                             DoubleTouchTracker.State.kicker_forbidden,
-                            lambda: not self.kicker_has_possession(),
-                            'kicker kicks or fumbles ball')
+                            # The ball is no longer in restart, we have begun playing
+                            (lambda: main.game_state().is_playing() or
+                             # We aren't in a restart anymore
+                             main.root_play().play is None or
+                             not main.root_play().play.__class__.is_restart()),
+                            'ball has moved and is now in play')
 
         self.add_transition(DoubleTouchTracker.State.kicker_forbidden,
                             DoubleTouchTracker.State.other_robot_touched,
                             lambda: self.other_robot_touching_ball(),
                             'another robot has touched the ball')
 
-    def kicker_has_possession(self):
-        if self.kicker_shell_id != None:
-            for bot in main.our_robots():
-                if bot.shell_id() == self.kicker_shell_id:
-                    # we use two methods here because the ball-sensor output is often jittery
-                    return bot.has_ball() or evaluation.ball.robot_has_ball(
-                        bot)
-        return False
+    def _has_ball(self, bot):
```

Comment:
Probably change to touching_ball. Hopefully our cookies are good.

Reply:
hopefully this can actually keep track of the ball and actually see which bot last touched it by seeing the acceleration but I wonder if that will happen anytime soon :P